### PR TITLE
JSDK-2985 - Cleaning up event listeners 

### PR DESCRIPTION
**Pull request by PikaJoyce**
JIRA Ticket : [JSDK-2985](https://issues.corp.twilio.com/browse/JSDK-2985)

This PR is for fixing the event listeners not being cleaned up when the participant is disconnected. This also includes a fix for a similar event emitter leak found on the `encodingParameters` object in peer to peer rooms. 

Completed:
- [x] Clean up event listeners on LocalParticipant
- [x] Validate that there are no more event emitter memory leaks in group and peer to peer rooms

Some manual tests I've tried:
- [x] Bob & Alice join p2p room : Alice turns on audio/video 10x times
- [x] Bob & Alice join p2p room : Alice disconnects/connects room 10x times (FIXED)
- [x] Bob & Alice join group room : Alice disconnects/connects room 10x times 
- [x] Bob & Alice join group room : Alice turns on audio/video 10x times

Please let me know if there are any other tests you'd like me to try!

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: lib/encodingparameters.js [hunk 1]
```diff
@@ -32,6 +32,9 @@ class EncodingParametersImpl extends EventEmitter {
       maxVideoBitrate: {
         value: encodingParameters.maxVideoBitrate,
         writable: true
+      },
+      count: {
+        value: listenerCount++
```
Comment (manjeshbhargav) [inline lib/encodingparameters.js:37]:
```suggestion
```

File: lib/localparticipant.js [hunk 2]
```diff
@@ -222,14 +222,25 @@ class LocalParticipant extends Participant {
         self.removeListener('trackDisabled', localTrackDisabled);
         self.removeListener('trackEnabled', localTrackEnabled);
         self.removeListener('trackStopped', localTrackStopped);
-
```
Comment (manjeshbhargav) [inline lib/localparticipant.js:225]: Please add this new line back.

File: lib/encodingparameters.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 'use strict';
 
 const EventEmitter = require('events').EventEmitter;
-
+let listenerCount = 0;
```
Comment (manjeshbhargav) [inline lib/encodingparameters.js:4]:
```suggestion
```

File: lib/localparticipant.js [hunk 4]
```diff
@@ -222,14 +222,25 @@ class LocalParticipant extends Participant {
...
+          if (trackEvents) {
+            Object.keys(trackEvents).forEach(event => {
+              const emitter = trackEvents[event];
+              if (typeof emitter === 'function') {
+                track.removeListener(event, emitter);
+              } else {
+                track.removeListener(event, emitter[0]);
+              }
+            });
+          }
```
Comment (manjeshbhargav) [inline lib/localparticipant.js:243]: Why is this necessary?

File: test/unit/spec/signaling/v2/peerconnection.js [hunk 5]
```diff
@@ -396,6 +396,10 @@ describe('PeerConnectionV2', () => {
           it('does not update the state', () => {
             assert.deepEqual(test.pcv2.getState(), before);
           });
+          it('removes event listeners on the encoding parameter', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/peerconnection.js:399]:
```suggestion
          it('removes the "changed" event listener on the underlying EncodingParameters', () => {
```

**Review by manjeshbhargav — CHANGES_REQUESTED** [review]

File: lib/localparticipant.js (hunk 4 above)
Comment (PikaJoyce) [reply lib/localparticipant.js:244]: I added this to remove the events on the track. However, not all of the callbacks are functions. Some appear to be an array. I'm essentially trying to get to **each** event on the track to remove them.

Comment (PikaJoyce) [reply lib/localparticipant.js:244]: Here's what it looks like 
image

I was also thinking about using [removeAllListeners](https://nodejs.org/api/events.html#events_emitter_removealllisteners_eventname).

Comment (manjeshbhargav) [reply lib/localparticipant.js:244]: After some debugging, we found that this is not necessary. Please remove this change.

**Review by manjeshbhargav — APPROVED** [review]

**Review by manjeshbhargav — APPROVED** [review]
